### DISCUSSION TITLE
반복 일정 관련 ERD 수정 논의

### DISCUSSION THREAD

**Discussion by jiwon83** [What-s-Your-Plan/wypl-backend#24] [General]
# 💢 문제 상황

**기존 반복 일정**

- 미래 시점에 대한 반복일정을 스케줄러를 통해 하루 전 생성
- 미래 시점에 대한 일정 데이터는 물리적으로 존재하지 않으므로, 정보 수정과 삭제가 모두 불가능하다.

**애플 캘린더 분석**

- 미래 일정 삭제 가능
![애플캘린더1](https://github.com/user-attachments/assets/38064087-6b4b-4e24-bd55-050b09105bbb)

- 미래 일정 수정 가능
![애플캘린더2](https://github.com/user-attachments/assets/39466d67-c8af-42d5-b934-a9a0e5f95f1e)

# 💦 해결 과정

|AS-IS | TO-BE|
| --- | --- |
|<img width="800" height="400" alt="image" src="https://github.com/user-attachments/assets/ca8f6d9a-90b6-4126-b35c-cc5c15785328"> | <img width="800" height="400" alt="image (1)" src="https://github.com/user-attachments/assets/b7730e82-fabd-4bb3-a5c3-d1acec61b2b7">| 

<br/>

> ex) 반복 주기의 수정
> 2024-09-01 ~ 2024-10-30인 일정에 대해, 2024-09-22일에 `매주 월요일` 반복 →  `매주 수요일` 반복으로 수정
> 
> Before
> 
> 2024-09-01 ~ 2024-10-30의 일정정보는 그대로, 해당 일정에 대한 반복주기 테이블에 수요일인 반복주기 레코드가 추가로 생성. 실제 일정 테이블은 스케줄러를 통해 하루 전 생성
> 
> After
> 
> 2024-09-01 ~ 2024-09-22 일정, 2024-09-23 ~ 2024-10-30의 일정으로 2개의 레코드로 분리

## 상세 변경 내용

AS-IS

- 반복 일정 생성 시, 반복주기 테이블로 반복 정보를 저장
- 반복 주기 변경 시, 변경된 반복주기로 새로운 레코드 생성

<img width="379" alt="image (2)" src="https://github.com/user-attachments/assets/13e0b9f9-df2f-4e39-b067-60f33cb785be">

TO-BE

- 반복 시작 일시, 반복 종료 일시 컬럼 추가
- 반복 일정 생성 시, 하나의 일정 테이블을 저장
- 반복 주기 변경 시, 새로운 반복주기에 대해 새로운 일정 레코드 생성

<img width="768" alt="image (3)" src="https://github.com/user-attachments/assets/cd492ec0-ef68-4509-b968-8e7c3520f923">

# ✅ 결과

- 미래 시점의 일정의 수정, 삭제 가능
- 스케줄러 개발 불필요

**Comment by KIMSEI1124** [comment]
👍🏼 최대한 자세히 작성해 주셔서 감사합니다!

미래의 데이터를 임의로 만든다는 것이 좋지 않은 방향성이라고 생각했는데 이렇게 해결하는 것을 보니 좋다고 생각이 드네요!
실제로 개발하면서 문제가 또 발생할 수도 있지만 현재로서는 가장 좋은 방향성인 것 같습니다!
